README.md
# correios_frete
[![Build Status](https://travis-ci.org/FelipeMourao/correios-frete.svg?branch=master)](https://travis-ci.org/FelipeMourao/correios-frete)
[![License: MIT](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://github.com/FelipeMourao/correios-frete/blob/master/LICENSE)
[![Codemagic build status](https://api.codemagic.io/apps/5e2af32b151bb60016f91e91/5e2af32b151bb60016f91e90/status_badge.svg)](https://codemagic.io/apps/5e2af32b151bb60016f91e91/5e2af32b151bb60016f91e90/latest_build)

Calcular o Preço e o Prazo da Entregas de encomendas. Este pacote utilizar API [http](https://pub.dev/packages/http) e [xml2json](https://pub.dev/packages/xml2json)

## Usage

```dart

dependencies:
     flutter:
        sdk: flutter
     correios_frete:

```


## Como calcular o valor do frete com Preço e Prazo

```dart

  // import 'package:correios_frete/correios_frete.dart';

  Result resultado = await CalcPriceTerm("0","0","04014","69097374","13480010","1","1","20","20","20",
      "0","N","N","N","xml","3");

    print("Valor do frete: ${resultado.valor}");
    print("Prazo de Entrega: ${resultado.prazo}");
    print("Valor da Mão Própria: ${resultado.vmp}");
    print("Valor do Aviso de Recebimento: ${resultado.valor}");
    print("Valor do Valor Declarado: ${resultado.vvd}");
    print("Possui Entrega Domiciliar (S - Sim , N - Não): ${resultado.entrDom}");
    print("Possui Entrega ao Sábados (S - Sim , N - Não): ${resultado.entrSa}");
    print("Valor sem Adicionais: ${resultado.vsa}");

```

## Documentação

Para obter ajuda sobre como funcionar o processo, consulte a [documentação](https://correios.com.br/enviar-e-receber/ferramentas/calculador-remoto-de-precos-e-prazos/pdf/manual-de-implementacao-do-calculo-remoto-de-precos-e-prazos/at_download/file)

CHANGELOG.md
## [0.0.1] - 2020-01-24.

* Initial Release: Implementation of the CalcPriceTerm and CalcTerm methods


test/correios_frete_test.dart
import 'package:correios_frete/correios_frete.dart';

void main() async {
  Result resultado = await CalcPriceTerm(
      "0",
      "0",
      "04014",
      "69097374",
      "13480010",
      "1",
      "1",
      "20",
      "20",
      "20",
      "0",
      "N",
      "N",
      "N",
      "xml",
      "3");

  print("Valor do frete: ${resultado.valor}");
  print("Prazo de Entrega: ${resultado.prazo}");
  print("Valor da Mão Própria: ${resultado.vmp}");
  print("Valor do Aviso de Recebimento: ${resultado.valor}");
  print("Valor do Valor Declarado: ${resultado.vvd}");
  print("Possui Entrega Domiciliar (S - Sim , N - Não): ${resultado.entrDom}");
  print("Possui Entrega ao Sábados (S - Sim , N - Não): ${resultado.entrSa}");
  print("Valor sem Adicionais: ${resultado.vsa}");

}

lib/correios_frete.dart
import 'dart:convert';
import 'dart:async';
import 'package:http/http.dart' as http;
import 'package:xml2json/xml2json.dart';


Future<Result> CalcPriceTerm(
    String nCdEmpresa,
    String sDsSenha,
    String nCdServico,
    String sCepOrigem,
    String sCepDestino,
    String nVlPeso,
    String nCdFormato,
    String nVlComprimento,
    String nVlAltura,
    String nVlLargura,
    String nVlDiametro,
    String sCdMaoPropria,
    String nVlValorDeclarado,
    String sCdAvisoRecebimento,
    String StrRetorno,
    String nIndicaCalculo) async {

  Xml2Json xml2json = new Xml2Json();

  if(nCdEmpresa == null){
    nCdEmpresa = "08082650";
  }

  if(sDsSenha == null){
    sDsSenha = "564321";
  }

  if(StrRetorno == null){
    StrRetorno = "xml";
  }

  if(nIndicaCalculo == null){
    nIndicaCalculo = "3";
  }

  if(nVlPeso == null){
    nVlPeso = "0";
  }

  if(nVlComprimento == null){
    nVlComprimento = "0";
  }

  if(nVlAltura == null){
    nVlAltura = "0";
  }

  if(nVlLargura == null){
    nVlLargura = "0";
  }

  if(nVlDiametro == null){
    nVlDiametro = "0";
  }

  if(nVlValorDeclarado == null){
    nVlValorDeclarado = "0";
  }

  try {

    http.Response reponse = await http.get(
        "http://ws.correios.com.br/calculador/CalcPrecoPrazo.asmx/CalcPrecoPrazo?nCdEmpresa=$nCdEmpresa&"
            "sDsSenha=$sDsSenha&nCdServico=$nCdServico&sCepOrigem=$sCepOrigem&sCepDestino=$sCepDestino&nVlPeso=$nVlPeso&nCdFormato=$nCdFormato&nVlComprimento=$nVlComprimento&"
            "nVlAltura=$nVlAltura&nVlLargura=$nVlLargura&nVlDiametro=$nVlDiametro&sCdMaoPropria=$sCdMaoPropria&nVlValorDeclarado=$nVlValorDeclarado&sCdAvisoRecebimento=$sCdAvisoRecebimento&StrRetorno=$StrRetorno&nIndicaCalculo=$nIndicaCalculo");

    if(reponse.statusCode==200){

      xml2json.parse(reponse.body);

      var resultadomap = xml2json.toGData();

      return Result.fromJson(json.decode(resultadomap)["cResultado"]["Servicos"]["cServico"]);

    } else if(reponse.statusCode==500) {

      print(reponse.body);

    }
  } catch (e) {
    print(e);
  }
}


Future<Result> CalcTerm(
    {String nCdServico,
      String sCepOrigem,
      String sCepDestino,
      String StrRetorno,
      String nIndicaCalculo}) async {

  Xml2Json xml2json = new Xml2Json();

  if(StrRetorno == null){
    StrRetorno = "xml";
  }

  if(nIndicaCalculo == null){
    nIndicaCalculo = "3";
  }

  try {

    http.Response reponse = await http.get(
        "http://ws.correios.com.br/calculador/CalcPrazo.asmx/CalcPrazo?&nCdServico=$nCdServico&sCepOrigem=$sCepOrigem&sCepDestino=$sCepDestino");

    if(reponse.statusCode==200){

      xml2json.parse(reponse.body);

      var resultadomap = xml2json.toGData();

      return Result.fromJson(json.decode(resultadomap)["cResultado"]["Servicos"]["cServico"]);
    }
  } catch (e) {
    print(e);
  }
}



class Result {
  Result(this.codigo, this.valor, this.prazo, this.vsa, this.vvd, this.vmp,
      this.entrDom, this.entrSa, this.error, this.msgErro, this.vRec);

  Result.fromJson(Map<String, dynamic> json)
      : codigo = json["Codigo"]["\$t"],
        valor = json["Valor"]["\$t"],
        prazo = json["PrazoEntrega"]["\$t"],
        vsa = json["ValorSemAdicionais"]["\$t"],
        vRec = json["ValorAvisoRecebimento"]["\$t"],
        vmp = json["ValorMaoPropria"]["\$t"],
        vvd = json["ValorValorDeclarado"]["\$t"],
        entrDom = json["EntregaDomiciliar"]["\$t"],
        entrSa = json["EntregaSabado"]["\$t"],
        error = json["Erro"]["\$t"],
        msgErro = json["MsgErro"]["\$t"];

  Map<String, dynamic> toJson() => {
        'Codigo': codigo,
        'Valor': valor,
        'PrazoEntrega': prazo,
        'ValorSemAdicionais': vsa,
        'ValorMaoPropria': vmp,
        'ValorValorDeclarado': vvd,
        'EntregaDomiciliar': entrDom,
        'EntregaSabado': entrSa,
        'Erro': error,
        'MsgErro': msgErro,
        'ValorAvisoRecebimento': vRec
      };

  final String codigo;

  final String valor;

  final String prazo;

  final String vsa;

  final String vmp;

  final String vvd;

  final String entrDom;

  final String entrSa;

  final String error;

  final String msgErro;

  final String vRec;
}


example/lib/screen/result.dart
import 'package:correios_frete/correios_frete.dart';
import 'package:flutter/material.dart';

class Resultado extends StatelessWidget {

  Map<String, dynamic> informacoes;

  Resultado(this.informacoes);

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      backgroundColor: Colors.lightBlue,
      appBar: AppBar(
        title: Text(
          "Resultado",
          style: TextStyle(color: Colors.white),
        ),
        backgroundColor: Colors.lightBlueAccent,
      ),
      body: Center(
        child: FutureBuilder<Result>(
          future: CalcPriceTerm(
              informacoes["nCdEmpresa"],
              informacoes["sDsSenha"],
              informacoes["nCdServico"],
              informacoes["sCepOrigem"],
              informacoes["sCepDestino"],
              informacoes["nVlPeso"],
              informacoes["nCdFormato"],
              informacoes["nVlComprimento"],
              informacoes["nVlAltura"],
              informacoes["nVlLargura"],
              informacoes["nVlDiametro"],
              informacoes["sCdMaoPropria"],
              informacoes["nVlValorDeclarado"],
              informacoes["sCdAvisoRecebimento"],
              informacoes["StrRetorno"],
              informacoes["nIndicaCalculo"]),
          builder: (context, snapshot) {
            switch (snapshot.connectionState) {
              case ConnectionState.none:
              case ConnectionState.waiting:
                return Container(
                  height: 200,
                  width: 200,
                  alignment: Alignment.center,
                  child: CircularProgressIndicator(
                    valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
                    strokeWidth: 5,
                  ),
                );
              default:
                if (snapshot.hasError)
                  return Container();
                else if (snapshot.data.error == "0") {
                  return Center(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.stretch,
                      children: <Widget>[
                        Text("Valor do Frete: R\$ ${snapshot.data.valor}",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                        Text("Prazo para Entrega: ${snapshot.data.prazo} dias",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                        Text(
                            "Preço Adicional do Serviço Mão Própria: R\$ ${snapshot.data.vmp}",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                        Text(
                            "Preço Adicional do Serviço Aviso de Recebimento: R\$ ${snapshot.data.vRec}",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                        Text(
                            "Preço Adicional do Serviço Valor Declarado: R\$ ${snapshot.data.vvd}",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                        Text(
                            "Possui Entrega Domiciliária?: ${snapshot.data.entrDom == "S" ? "Sim" : "Não"}",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                        Text(
                            "Possui Entrega aos Sábados?: ${snapshot.data.entrSa == "S" ? "Sim" : "Não"}",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                        Text("Valor Sem Adicionais: R\$ ${snapshot.data.vsa}",
                            style: TextStyle(color: Colors.white, fontSize: 25)),
                      ],
                    ),
                  );
                } else {
                  return Center(
                    child: Column(
                      children: <Widget>[
                        Text(
                            snapshot.data.error,
                            style: TextStyle(color: Colors.red, fontSize: 25)),
                        Text(snapshot.data.msgErro,
                            style: TextStyle(color: Colors.red, fontSize: 25)),
                      ],
                    ),
                  );
                }
            }
          },
        ),
      ),
    );
  }
}


example/lib/screen/teste_frete.dart
import 'package:example/screen/result.dart';
import 'package:flutter/material.dart';

class TesteFrete extends StatefulWidget {
  @override
  _TesteFreteState createState() => _TesteFreteState();
}

class _TesteFreteState extends State<TesteFrete> {
  Map<String, dynamic> parametros = Map();
  bool _valueMaePropria = false;
  bool _valueAvisoRecebimento = false;
  var _currencies = ['SEDEX', 'PAC', 'SEDEX 10', 'SEDEX 12', 'SEDEX Hoje'];

  @override
  void initState() {
    super.initState();
    parametros["nCdServico"] = "04014";
  }
  var _currenciesInitial = "SEDEX";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.lightBlue,
      appBar: AppBar(
        title: Text(
          "Teste Frete",
          style: TextStyle(color: Colors.white),
        ),
        backgroundColor: Colors.lightBlueAccent,
      ),
      body: SingleChildScrollView(
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              SizedBox(height: 40),
              TextField(
                decoration: InputDecoration(
                    labelText: "CÓDIGO ADMINISTRATIVO",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["nCdEmpresa"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "SENHA DE ACESSO",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["sDsSenha"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "CEP DE ORIGEM SEM HÍFEN",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["sCepOrigem"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "CEP DO DESTINO SEM HÍFEN",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["sCepDestino"] = text;
                },
              ),
              Divider(),
              Row(
                children: <Widget>[
                  RaisedButton(
                    color: parametros["nCdFormato"] == "1"
                        ? Colors.blue[700]
                        : Colors.blueAccent,
                    child: Text("Caixa/Pacote",
                        style: TextStyle(color: Colors.white)),
                    onPressed: () {
                      setState(() {
                        parametros["nCdFormato"] = "1";
                      });
                    },
                  ),
                  SizedBox(width: 30),
                  RaisedButton(
                    color: parametros["nCdFormato"] == "2"
                        ? Colors.blue[700]
                        : Colors.blueAccent,
                    child: Text("Rolo/Prisma",
                        style: TextStyle(color: Colors.white)),
                    onPressed: () {
                      setState(() {
                        parametros["nCdFormato"] = "2";
                      });
                    },
                  ),
                  SizedBox(width: 30),
                  RaisedButton(
                    color: parametros["nCdFormato"] == "3"
                        ? Colors.blue[700]
                        : Colors.blueAccent,
                    child:
                        Text("Envelope", style: TextStyle(color: Colors.white)),
                    onPressed: () {
                      setState(() {
                        parametros["nCdFormato"] = "3";
                      });
                    },
                  )
                ],
              ),
              Divider(),
              DropdownButton<String>(
                items: _currencies.map((String codigo) {
                  return DropdownMenuItem<String>(
                    value: codigo,
                    child: Text(codigo),
                  );
                }).toList(),
                onChanged: (value) {
                  codigoServico(value);
                  print(parametros["nCdServico"]);
                },
                value: _currenciesInitial,
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "PESO",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["nVlPeso"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "COMPRIMENTO",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["nVlComprimento"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "ALTURA",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["nVlAltura"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "LARGURA",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["nVlLargura"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "DIÂMETRO",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["nVlDiametro"] = text;
                },
              ),
              Divider(),
              TextField(
                decoration: InputDecoration(
                    labelText: "VALOR DECLARADO",
                    labelStyle: TextStyle(color: Colors.white),
                    border: OutlineInputBorder()),
                style: TextStyle(color: Colors.white, fontSize: 25),
                keyboardType: TextInputType.number,
                onChanged: (text) {
                  parametros["nVlValorDeclarado"] = text;
                },
              ),
              Divider(),
              SwitchListTile(
                title:
                    Text("Mão Própria?", style: TextStyle(color: Colors.white)),
                activeColor: Colors.green,
                value: _valueMaePropria,
                onChanged: (value) {
                  value == true
                      ? parametros["nVlValorDeclarado"] = "S"
                      : parametros["nVlValorDeclarado"] = "N";
                },
              ),
              Divider(),
              SwitchListTile(
                title: Text("Aviso de Recebimento?",
                    style: TextStyle(color: Colors.white)),
                activeColor: Colors.green,
                value: _valueAvisoRecebimento,
                onChanged: (value) {
                  value == true
                      ? parametros["sCdAvisoRecebimento"] = "S"
                      : parametros["sCdAvisoRecebimento"] = "N";
                },
              ),
              Divider(),
              RaisedButton(
                color: Colors.blueAccent,
                child: Text("Buscar", style: TextStyle(color: Colors.white)),
                onPressed: () {
                  Navigator.push(
                      context,
                      MaterialPageRoute(
                          builder: (context) => Resultado(parametros)));
                },
              ),
            ],
          ),
        ),
      ),
    );
  }

  void codigoServico(String codigo) {
    switch (codigo) {
      case "SEDEX":
        setState(() {
          this._currenciesInitial = codigo;
          parametros["nCdServico"] = "04014";
        });
        break;
      case "PAC":
        setState(() {
          this._currenciesInitial = codigo;
          parametros["nCdServico"] = "04510";
        });
        break;
      case "SEDEX 10":
        setState(() {
          this._currenciesInitial = codigo;
          parametros["nCdServico"] = "04782";
        });
        break;
      case "SEDEX 12":
        setState(() {
          this._currenciesInitial = codigo;
          parametros["nCdServico"] = "04790";
        });
        break;
      case "SEDEX Hoje":
        setState(() {
          this._currenciesInitial = codigo;
          parametros["nCdServico"] = "04804";
        });
        break;
    }
  }
}
